### ISSUE TITLE
Update to OpenSSL 3.0.13 (& 1.1.1w) in our binary release build process.

### ISSUE THREAD

**Issue by gpshead** [python/cpython#109991]
# Bug report

### Bug description:

We need to upgrade the OpenSSL versions we build & bundle into our binary releases before the next release. More security fixes  as usual. In particular https://nvd.nist.gov/vuln/detail/CVE-2023-4807 applies to our 64-bit Windows binaries.

Pick the latest 3.0.x and 1.1.1 releases at the time the work is done.  3.0.11 today, and if we build binaries for older shipping-with-1.1 branches, 1.1.1w. We should update the binary _build_ tooling in older release branches for those to at least reference and pull in 1.1.1w even if we aren't shipping new binary releases on those ourselves.

### CPython versions tested on:

3.8, 3.9, 3.10, 3.11, 3.12

### Operating systems tested on:

macOS, Windows

### Linked PRs
* gh-110002
* gh-110003
* gh-110004
* gh-110005
* gh-110006
* gh-110007
* gh-110008
* gh-110009
* gh-110010
* gh-110054
* gh-110055
* gh-110056
* gh-110059
* gh-110090
* gh-111265
* gh-111266
* gh-115043
* gh-115047
* gh-115048
* gh-115050
* gh-115052
* gh-115053
* gh-115054
* gh-115055
* gh-115057

**Comment by zooba** [comment]
I just pushed updated Windows builds of OpenSSL to the cpython-bin-deps for OpenSSL 1.1.1w and 3.0.11.

**Comment by ned-deily** [comment]
PRs submitted and/or merged to update OpenSSL in GHA CI for main/3.13, 3.12, 3.11, 3.10, 3.9, and 3.9.
PRs submitted and/or merged to update to OpenSSL 3.0.11 in macOS installer builds for main, 3.12, and 3.11.

**Comment by Yhg1s** [comment]
Are all the necessary changes in for 3.12 to be released? (I think they are, just double-checking.)

**Comment by ned-deily** [comment]
Yes, they are now.

**Comment by v-leejosh** [comment]
Can an update use OpenSSL 3.0.12 be done? https://nvd.nist.gov/vuln/detail/CVE-2023-5363 affects 3.0.11, and has been fixed in 3.0.12 https://www.openssl.org/news/vulnerabilities.html

**Comment by zware** [comment]
The next round of updates will be handled in a new issue, which you can feel free to open.  However, note that there's already a 3.0.13 pending with another low-severity fix, and it's not immediately obvious to me that Python is affected by either one (but I am not an expert :) ).  Unless it can be shown that Python *is* affected by CVE-2023-5363, we may prefer to wait until 3.0.13 is available before updating.

---

@ambv, could you take a look at the pending 3.8 and 3.9 PRs so we can close this issue?

**Comment by zooba** [comment]
> Unless it can be shown that Python _is_ affected by [CVE-2023-5363](https://github.com/advisories/GHSA-xw78-pcr6-wrg8), we may prefer to wait until 3.0.13 is available before updating

Doesn't look like it'd impact us. We don't expose the encryption algorithms, and the TLS implementation is unaffected.

That said, a CVE that tags a version of OpenSSL that we include in our installer is effectively treated as a blocking vulnerability for our release by more and more environments these days. As much as we can argue that they're wrong, it may prevent people from being able to use Python at all, and will definitely convince some people that we (the core/release team) "don't understand security" (yes, direct quote), so for the sake of public perception, we should try to include the latest version in our next release even if it doesn't actually make any difference.

**Comment by gpshead** [comment]
https://nvd.nist.gov/vuln/detail/CVE-2023-5678 is the other issue. OpenSSL rightfully [calls it "Low" severity](https://www.openssl.org/news/secadv/20231106.txt). That seems accurate. The NVD has a misleading for most applications "High" score stamped on it.

Modules/_ssl.c (and hashlib) do not call any of the `DH_*` or `EVP_PKEY_*` APIs listed in that CVE.  **SSL/TLS is not impacted.**  This is also not a reason to expedite any releases.  The OpenSSL project has not yet finalized their fix for that one - when they do it'll become 3.0.13.

**Comment by ned-deily** [comment]
OpenSSL 3.0.13 was released on 2024-01-30. There are no further 1.1.1 releases planned past 1.1.1w.

**Comment by zooba** [comment]
A new Windows build of 3.0.13 is published now, so we can update the tags and validate (I have another PR checked out so can't do it immediately)

**Comment by pablogsal** [comment]
This is marked as release-blocker with the 3.11 tag. I can see a bunch of 3.11 backports but the issue it's still open. I am holding the next 3.11 bugfix release until tomorrow at least to know if anything is missing here before we can go with the release.

**Comment by zooba** [comment]
This issue has been used for a few OpenSSL updates, which is why there are so many PRs and backports. After 3.0.13 is done, we should close this and require a new one for next time.

**Comment by pablogsal** [comment]
So just to confirm explicitly: should I hold the 3.11 release on this or not?

**Comment by ned-deily** [comment]
> should I hold the 3.11 release on this or not?

At this point, probably the better to do so. I'll update the macOS installer in a PR. And then we also should have a PR to update the CI runs and Tools/ssl/multissltests.py.

**Comment by Yhg1s** [comment]
Is there anything left to do before the next 3.12 release (scheduled for today)?

**Comment by ned-deily** [comment]
> Is there anything left to do before the next 3.12 release (scheduled for today)?

I think we are good to go for 3.12 and 3.11. There could be backports needed for 3.10, 3.9, and 3.8 at the discretion of their release managers.

**Comment by zware** [comment]
The 3.8-3.10 Windows builds are still on 1.1.1w and would require a larger-than-usual backport to jump up to 3.0.  I'm not sure it's worthwhile since we're no longer producing binaries for those versions.  If we do decide to do that backport I think it's worth a new issue, so I'm closing this one.
